accepted/2024/observable_collection.md

```diff
@@ -0,0 +1,205 @@
+  - Pro: Core API is simple while providing an affordance to the user to solve
+    the problem without requiring changes from the UI framework.
+  - Con: More complex state, might not play well with WPF's `CollectionView`.
+
+> [!NOTE]
+> Doesn't seem feasible. Handlers of the event likely assume the collection
+> was just changed to include the one modification. Just translating the
+> event would mean they get a series of events but the handler sees the
+> final collection in all invocations.
+
+* **Event enumerator**. We could expose a method on
+  `NotifyCollectionChangedEventArgs` that returns an
+  `IEnumerable<NotifyCollectionChangedEventArgs>`. For single item events it
+  returns itself, otherwise it creates a new event arg per item. This requires
+  UI controls to make code changes if they don't support bulk notifications, but
+  the change is fairly targeted.
+  - Pro: Simple API
+  - Con: Still breaks consumers, just provides an easier path to adapt to the
+    new behavior without having to fully support bulk notifications.
+
+> [!NOTE]
+> Doesn't seem feasible. Handlers of the event likely assume the collection
+> was just changed to include the one modification. Just translating the
+> event would mean they get a series of events but the handler sees the
+> final collection in all invocations.
+
+* **Handling it at the CollectionView level**. WPF seems to have an indirection
+  between the data-bound UI controls and the observable collections via the
+  collection view. Maybe we could expose a property on `CollectionView` that
+  controls whether the collection view will translate bulk events to individual
+  events.
+  - Pro: Simpler API for the base, with an opt-out switch at the level of WPF.
+  - Con: Needs more investigation to determine feasibility, would only be helpful for WPF
+
+* **Separate API**. We could add a new interface
+  `IBatchedNotifyCollectionChanged` interface that `ObservableCollection<T>`
+  also implements. The interface adds new even `BulkCollectionChanged`.
+  - Pro: Consumers opt-into bulk events by subscribing to a different event.
+    This allows a single instance of `ObservableCollection<T>` to be bound to
+    multiple UI elements and leverage bulk updates when it's supported while
+    falling back when it's not. This would also allow change of semantics to
+    allow for non-contiguous updates, for instance.
+  - Con: Complicates that API; requires all consumers to explicitly opt-in, even
+    if happen to already support bulk notifications. Would also likely require
+    adding a corresponding interface to WinUI that .NET can project to. Will
+    likely take longer to be supported by UI frameworks.
+
+> [!NOTE]
+> For the same reason event adapters aren't viable just having a new
+> interface is probably not sufficient. Rather, we'd need a different
+> implementation like `BulkObservableCollection<T>`. However, for WinUI
+> we likely also still need an interface too.
```

Comment:
This would not only impact WinUI (3), but UWP as well. In general, it would impact anyone trying to ever marshal this new interface across the WinRT boundary. The fact it impacts WinUI 3 and UWP is just a consequence of this.